crates/bevy_core_pipeline/src/taa/taa.wgsl

```diff
@@ -0,0 +1,179 @@
+        closest_depth = d_bl;
+    }
+    if d_br > closest_depth {
+        closest_velocity = v_br;
+    }
+    let previous_uv = uv - closest_velocity;
+
+    // Reproject to find the equivalent sample from the past
+    // Uses 5-sample Catmull-Rom filtering (reduces blurriness)
+    // https://gist.github.com/TheRealMJP/c83b8c0f46b63f3a88a5986f4fa982b1
+    // https://vec3.ca/bicubic-filtering-in-fewer-taps
+    // https://developer.nvidia.com/gpugems/gpugems2/part-iii-high-quality-rendering/chapter-20-fast-third-order-texture-filtering
+    // https://www.activision.com/cdn/research/Dynamic_Temporal_Antialiasing_and_Upsampling_in_Call_of_Duty_v4.pdf#page=68
+    let sample_position = previous_uv * texture_size;
+    let texel_center = floor(sample_position - 0.5) + 0.5;
+    let f = sample_position - texel_center;
+    let w0 = f * (-0.5 + f * (1.0 - 0.5 * f));
+    let w1 = 1.0 + f * f * (-2.5 + 1.5 * f);
+    let w2 = f * (0.5 + f * (2.0 - 1.5 * f));
+    let w3 = f * f * (-0.5 + 0.5 * f);
+    let w12 = w1 + w2;
+    let texel_position_0 = (texel_center - 1.0) * texel_size;
+    let texel_position_3 = (texel_center + 2.0) * texel_size;
+    let texel_position_12 = (texel_center + (w2 / w12)) * texel_size;
+    var previous_color = vec3(0.0);
+    previous_color += sample_history(texel_position_12.x, texel_position_0.y) * w12.x * w0.y;
+    previous_color += sample_history(texel_position_0.x, texel_position_12.y) * w0.x * w12.y;
+    previous_color += sample_history(texel_position_12.x, texel_position_12.y) * w12.x * w12.y;
+    previous_color += sample_history(texel_position_3.x, texel_position_12.y) * w3.x * w12.y;
+    previous_color += sample_history(texel_position_12.x, texel_position_3.y) * w12.x * w3.y;
+
+    // Constrain past sample with 3x3 YCoCg variance clipping (reduces ghosting)
+    // YCoCg: https://advances.realtimerendering.com/s2014/index.html#_HIGH-QUALITY_TEMPORAL_SUPERSAMPLING, slide 33
+    // Variance clipping: https://developer.download.nvidia.com/gameworks/events/GDC2016/msalvi_temporal_supersampling.pdf
+    let s_tl = sample_view_target(uv + vec2(-texel_size.x, texel_size.y));
+    let s_tm = sample_view_target(uv + vec2(0.0, texel_size.y));
+    let s_tr = sample_view_target(uv + texel_size);
+    let s_ml = sample_view_target(uv - vec2(texel_size.x, 0.0));
+    let s_mm = RGB_to_YCoCg(current_color);
+    let s_mr = sample_view_target(uv + vec2(texel_size.x, 0.0));
+    let s_bl = sample_view_target(uv - texel_size);
+    let s_bm = sample_view_target(uv - vec2(0.0, texel_size.y));
+    let s_br = sample_view_target(uv + vec2(texel_size.x, -texel_size.y));
+    let moment_1 = s_tl + s_tm + s_tr + s_ml + s_mm + s_mr + s_bl + s_bm + s_br;
+    let moment_2 = (s_tl * s_tl) + (s_tm * s_tm) + (s_tr * s_tr) + (s_ml * s_ml) + (s_mm * s_mm) + (s_mr * s_mr) + (s_bl * s_bl) + (s_bm * s_bm) + (s_br * s_br);
+    let mean = moment_1 / 9.0;
+    let variance = sqrt((moment_2 / 9.0) - (mean * mean));
```

Comment:
nit, but I think this is stddev (sigma) rather than variance (sigma^2).

Reply:
Good point